fcm/fcm.go

```diff
@@ -0,0 +1,200 @@
+package fcm
+
+import (
+	"bufio"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+)
+
+import (
+	"github.com/google/cloud-print-connector/log"
+	"github.com/google/cloud-print-connector/notification"
+)
+
+const (
+	gcpFcmSubscribePath = "fcm/subscribe"
+)
+
+type FCM struct {
+	fcmServerBindURL string
+	cachedToken      string
+	tokenRefreshTime time.Time
+	clientID         string
+	proxyName        string
+	fcmTTLSecs       float64
+	FcmSubscribe     func(string) (interface{}, error)
+
+	notifications chan<- notification.PrinterNotification
+	dead          chan struct{}
+
+	quit chan struct{}
+}
+
+type Notification struct {
+	Data   [][]interface{}
+	Number int
+}
+
+type Data struct {
+	Notification string `json:"notification"`
+	Subtype      string `json:"subtype"`
+}
+
+type FcmMessage struct {
+	Category    string `json:"category"`
+	CollapseKey string `json:"collapse_key"`
+	Data        Data   `json:"data"`
+	From        string `json:"from"`
+	MessageID   string `json:"message_id"`
+	TimeToLive  int    `json:"time_to_live"`
+}
+
+func NewFCM(clientID string, proxyName string, fcmServerBindURL string, FcmSubscribe func(string) (interface{}, error), notifications chan<- notification.PrinterNotification) (*FCM, error) {
+	f := FCM{
+		fcmServerBindURL,
+		"",
+		time.Time{},
+		clientID,
+		proxyName,
+		0,
+		FcmSubscribe,
+		notifications,
+		make(chan struct{}),
+		make(chan struct{}),
+	}
+	return &f, nil
+}
+
+//  get token from GCP and connect to FCM.
+func (f *FCM) Init() {
+	iidToken := f.GetToken()
+	if err := f.ConnectToFcm(f.notifications, iidToken, f.dead, f.quit); err != nil {
+		for err != nil {
+			log.Errorf("FCM restart failed, will try again in 10s: %s", err)
+			time.Sleep(10 * time.Second)
+			err = f.ConnectToFcm(f.notifications, iidToken, f.dead, f.quit)
```

Comment:
what if failed, same as below.

Reply:
Retry happens every 10 second till err is nil. 
This is similar to what we do in xmpp.

Reply:
i mean it keeps repeating on failure without proceeding and connector is never up in this case? then what's the difference between this and aborting on failure?

Reply:
connector can run in 3 modes 
1) local
2) gcp (cloud)
3) both local and cloud

FCM.init is executed only when cloud is enabled and it is the last step while connector is coming up. Yes connector keeps retrying for a successful connection  with notification server (either xmpp or fcm) every failure attempt (happens very rarely) will be logged to log file.  

here are possible causes for failure:
1) google service down (it will not be down for ever).
2) n/w issues 
   intermittent n/w issues are automatically taken care if issue persist based log messages user should be able to fix the issue.

We chose infinite retry because this will be in consistent with the current xmpp retries ans second if we give up after say 3 reties it will keep connector up (in wrong state), connector in this state will not receive new print job notifications and may take very long time before user realizes this.